File: 22-括号生成.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        string s;
        vector<string> ans;
        backtrack(ans,0,0,s,n);
        return ans;
    }
    void backtrack(vector<string>& ans, int left, int right, string& s, int n){
        if(s.size() == 2*n)
            ans.push_back(s);
        else{
            if(left<n){
                s.push_back('(');
                backtrack(ans,left+1,right,s,n);
                s.pop_back();
            }
            if(right<left){
                s.push_back(')');
                backtrack(ans,left,right+1,s,n);
                s.pop_back();
            }
        }
        return;
    }
};

File: 22-括号生成.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n == 1:
            return list({'()'})#借助set的去重功能，解决了重复问题，取巧了属于是
        res = set()
        for s in self.generateParenthesis(n-1):
            for j in range(0,len(s)+2):
                res.add(s[0:j]+'()'+s[j:])
        return list(res)